[오목 1,2단계] 다이스 미션 제출합니다.

moondev03 opened the pull request:
말리빈, 안녕하세요!!
안드로이드 7기 다이스🎲 입니다! 👋
이번 오목 미션 리뷰 잘 부탁드립니다 :)

<br>

# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

<br>

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

### OmokAdapter
5기 안드로이드 크루의 코드를 참고하여 금수 판단하는 코드 작성했습니다!
1단계에서 구현한 도메인 모델과 가져온 코드를 호환 가능하게 해주기 위해 `OmokAdapter`를 작성했습니다.

<br>

## 코드 리뷰 커뮤니케이션

<!--
리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
-->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

<br>

## 참고 자료

<!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

<br>

## 스크린샷

<!-- 필요하다면 스크린샷을 첨부하세요. -->

<img width="405" alt="image" src="https://github.com/user-attachments/assets/cc6db5f7-5f0a-48b3-87e2-ffdc0a269d17" />

etama123 committed:
docs: 기능 목록 작성

etama123 committed:
feat: 오목돌은 색과 위치를 가진다

etama123 committed:
feat: 오목판은 착수된 돌의 목록을 가진다

etama123 committed:
chore: klint 수정

etama123 committed:
refactor: Board class로 전환

etama123 committed:
feat: 오목판 돌 추가 기능 구현

etama123 committed:
feat(Point): 각 칸은 본인의 상태를 가진다

etama123 committed:
feat(Points): 오목판 칸 생성 기능 구현

etama123 committed:
feat(Board): 오목판 돌 추가 기능 구현

etama123 committed:
feat: 직전 돌 저장 기능 구현

moondev03 committed:
refactor: Board 개선

- Points 제거

moondev03 committed:
feat: 오목 승패 조건 판단 기능 구현

moondev03 committed:
refactor: 패키지 구조 변경

moondev03 committed:
test: 반대 색상을 반환하는 기능 테스트

moondev03 committed:
fix: FiveInRow 조건 수정

- 초기 count = 0

moondev03 committed:
feat: Omok View Interface & Impl 작성

moondev03 committed:
feat: 오목 게임 객체 생성

moondev03 committed:
test: 렌주룰 Board Fixture 추가

moondev03 committed:
feat: 3-3, 4-4 금수 판단 객체 추가

moondev03 committed:
feat: 도메인 모델 -> lib를 위한 adapter 객체 추가

moondev03 committed:
feat: 장목 금수 판단 객체 추가

moondev03 committed:
feat: 금수 판단 객체 추가

moondev03 committed:
fix: 착수 위치 검증 로직 수정

moondev03 committed:
test: 테스트용 보드 fixture 추가

moondev03 committed:
refactor: Place 성공 핸들러 추가

- 금수 판단

moondev03 committed:
refactor: 금수 판단 클래스를 Object로 변경

moondev03 committed:
refactor: PointState에서 Close 상태 제거

moondev03 committed:
feat: 착수 위치 재입력 기능 구현

moondev03 committed:
refactor: 불필요한 파일 제거

moondev03 committed:
refactor: 패키지 구조 변경

moondev03 committed:
refactor: 코드 가독성 개선

moondev03 committed:
refactor: 금수 판단 클래스 Object로 변경

moondev03 committed:
refactor: 객체 네이밍 수정

moondev03 committed:
refactor: 보드 크기 상수 사용하도록 수정

moondev03 committed:
refactor: findPoint 함수 개선

- 테스트 코드 추가

etama123 committed:
docs: 기능 목록 수정

malibinYun commented on src/main/kotlin/omok/model/OmokGame.kt:
```diff
@@ -0,0 +1,88 @@
+                    handlePlaceSuccess(result, board)
+                    return@retryOnException
+                }
+                is PlaceStoneResult.AlreadyPlaced -> throw IllegalArgumentException(ALREADY_PLACED_ERROR_MESSAGE)
+                is PlaceStoneResult.Closed -> throw IllegalArgumentException(CLOSED_ERROR_MESSAGE)
```
수를 둘 때마다 나타날 수 있는 게임의 상태를 sealed로 잘 표현해주셨네요 👍
그렇게 표현해두고 다시 throw를 하는 이유는 무엇인가요?
throw하지 않고, 해당 sealed 의 분기대로 코드 로직을 구성해보는 건 어떨까요?

malibinYun replied:
> 수를 둘 때마다 나타날 수 있는 게임의 상태를 sealed로 잘 표현해주셨네요 👍
그렇게 표현해두고 다시 throw를 하는 이유는 무엇인가요?

다시 throw 하셨던 이유가 궁금하네요~

moondev03 replied:
단순히 에러 메시지 출력 후 재시도를 위해 예외를 던져
`retryOnException`을 사용하고 싶었던 것 같습니다.

지금은 예외를 던지는 대신 각 결과에 맞게 처리될 수 있도록 수정했습니다!

malibinYun commented on src/main/kotlin/omok/model/OmokGame.kt:
```diff
@@ -0,0 +1,88 @@
+        outputView.printBoardStatus(board)
+    }
+
+    private fun judgeRule(
+        omokCountRule: OmokCountRule,
+        board: Board,
+    ): Boolean {
+        if (previousPoint == null) return true
+        return !omokCountRule.calculate(board, previousPoint!!)
+    }
```
게임 자체가 5개의 돌이 연속적으로 놓여있는지 보드의 값을 꺼내 비교해볼 수도 있겠어요.

게임 보드 자체가 5개 돌이 놓여 게임이 끝났다는 것을 판단할 수 있도록 만들어보는 건 어떨까요?
그 결과를 반환해주는 것도 어떤 상태를 나타내주는 객체에 역할을 맡겨볼 수 있겠네요 :)

malibinYun replied:
여기는 반영하지 않는 부분일까요?

moondev03 replied:
보드는 바둑판을 사용하는 다른 게임에서도 활용할 수 있도록 하고 싶었습니다.
때문에 위반, 승리 조건 등을 관리하는 RuleValidator를 보드에서 주입받도록 했습니다.

malibinYun commented on src/main/kotlin/omok/model/StoneColor.kt:
```diff
@@ -0,0 +1,14 @@
+    ;
+
+    companion object {
+        fun StoneColor.next(): StoneColor {
+            if (this == WHITE) return BLACK
+            return WHITE
+        }
```
코틀린의 if는 "문"이 아니라 "식"이어서
키워드를 붙여 쓸 수도 있답니다.
return if () ~~
요렇게 구성해볼 수 있어요 :)

---

그런데 StoneColor의 확장함수 형태로 구성한 큰 이유가 있을까요?_?

moondev03 replied:
enum 안에 작성해야 하는데 무의식적으로 companion object에 작성하게 되어 확장 함수로 작성했네요..

malibinYun commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,48 @@
+import omok.model.StoneColor
+import omok.model.rule.ForbiddenMoveJudge
+
+class Board {
+    val points: List<Point> =
+        (BOARD_MIN_SIZE..BOARD_MAX_SIZE).flatMap { row ->
+            (BOARD_MIN_SIZE..BOARD_MAX_SIZE).map { col ->
+                Point(Position(row, col))
+            }
+        }
```
보드판의 좌표들을 List가 아니라 다른 방식으로 표현해보는 건 어떨까요?
하나의 열을 하나의 클래스로 만들어볼 수도 있겠어요.
혹은 다른 자료구조를 활용하거나요.
또는 다른 좋은 방법을 떠올려보세요!

moondev03 replied:
Map<Point, PointState> 형태로 변경하고 BoardPoints라는 객체를 만들어 별도로 관리할 수 있게 수정했습니다!

malibinYun commented on src/main/kotlin/omok/model/board/Point.kt:
```diff
@@ -0,0 +1,20 @@
+package omok.model.board
+
+import omok.model.StoneColor
+
+data class Point(val position: Position) {
```
Point와 Position이 어떻게 다른지 다른 개발자들은 이름만 보고 한 번에 유추할 수 있을까요?

moondev03 replied:
Point 안에 State를 포함시키려 해서 발생한 문제였던 것 같습니다. 
리팩토링을 진행하면서 State를 더 이상 Point 안에서 관리하지 않기로 했고, 
그 과정에서 Position을 제거하고 Point에 역할을 부여했습니다!

malibinYun commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,48 @@
+            }
+        }
+
+    fun findPoint(position: Position): Point {
+        return requireNotNull(points.find { it.position == position }) { NOT_FOUND_POINT_ERROR_MESSAGE }
+    }
```
입력 좌표로 값을 찾지 못한 것은 "논리적인 오류"인가요?

moondev03 replied:
크루원들과 해당 내용에 대해서 `오류`라는 단어를 `모순`으로 생각해보라는 얘기가 나왔었습니다.

금수 규칙에 걸리더라도 바둑판 안에 돌을 착수한다면 모순이 아니니 논리적 오류이고,
바둑판 밖에 착수한다는 것은 모순이니 비논리적 오류이다.

이렇게 생각해도 괜찮을까요?

근데, 논리적 오류의 정의를 찾아보다
`비록 즉시 인식되지는 않지만 의도치 않은 또는 바라지 않은 결과나 다른 행동을 유발한다.`라는 내용이 있었습니다.

바둑판 범위 밖에 착수한다는 것도 충분히 개발자의 의도하에 처리된다면 논리적 오류로 볼 수 있지 않을까라는 생각이 들어서
논리적/비논리적 오류를 구분하는데 헷갈리는 부분이 아직 남아있는 것 같습니다.

우선은 null로 반환하도록 한 뒤
sealed class를 통해 처리되도록 해봤습니다!

malibinYun replied:
음~ 여기는 착수하는 게 아니라 값을 찾는 메서드인데요,
맥락이 같은 게 맞을까요?

또, 오류를 모순이라고 이름을 바꿨는데, "모순이 아니니 논리적 `오류`" 이다 라는 말은.. 모순이 아니니 논리적 모순이다 라고 말해야하는 게 모순되지 않은 상황이 아닐까~ 하는데요

잘못된 수를 두어서 exception이 발생했는데요, 이것은 예측 불가능한 "오류" 인가요?

malibinYun commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,48 @@
+    private fun handlePlaceSuccess(
+        color: StoneColor,
+        point: Point,
+    ): PlaceStoneResult {
+        if (color == StoneColor.BLACK && !ForbiddenMoveJudge.validate(this, point)) {
```
오목의 규칙이 게임마다 바뀌어야할 때, 이 형태는 유연하게 대처할 수 있는 형태일까요?

moondev03 replied:
오목 규칙을 관리하는 객체를 추가해서 모두 Board에서 처리하고 결과만 반환받도록 수정했습니다.

[358fd40](https://github.com/woowacourse/kotlin-omok/pull/108/commits/358fd40035cf23d9535329b55c5247303551ceea)

malibinYun commented on src/main/kotlin/omok/model/board/Position.kt:
```diff
@@ -0,0 +1,11 @@
+package omok.model.board
+
+data class Position(val x: Int, val y: Int) {
+    init {
+        require(listOf(x, y).all { it in Board.BOARD_MIN_SIZE..Board.BOARD_MAX_SIZE }) { INVALID_POSITION_MESSAGE }
+    }
```
좌표 자체가 오목판에 대한 관심을 가질 필요가 있을지 고민해보세요 :)

moondev03 replied:
좌표는 오목판 크기를 알 필요가 없다고 생각해서 검증 처리를 제거했습니다.

malibinYun commented on src/main/kotlin/omok/model/rule/OmokAdapter.kt:
```diff
@@ -0,0 +1,26 @@
+import omok.model.board.PointState
+import omok.model.rule.lib.ForbiddenMoveRule
+
+object OmokAdapter {
+    fun adaptOmokBoard(board: Board): List<List<Int>> {
+        val adapted = MutableList(Board.BOARD_MAX_SIZE) { MutableList(Board.BOARD_MAX_SIZE) { ForbiddenMoveRule.EMPTY_STONE } }
+
+        board.points.forEach { point ->
+            adapted[point.position.y - 1][point.position.x - 1] =
+                when (board.findPoint(point.position).state) {
+                    PointState.BLACK -> ForbiddenMoveRule.BLACK_STONE
+                    PointState.WHITE -> ForbiddenMoveRule.WHITE_STONE
+                    else -> ForbiddenMoveRule.EMPTY_STONE
+                }
+        }
+        return adapted
+    }
+
+    fun adaptOmokPoint(point: Point): Pair<Int, Int> {
+        return Pair(point.position.x - 1, point.position.y - 1)
+    }
+}
```
이 객체는 어떤 어뎁터 패턴을 띄기 보다, 변환기의 성격이 강해보이네요.

moondev03 replied:
Adapter 패턴을 잘 알아보지 않았던 것 같아요 ㅠㅠ
Converter로 네이밍을 수정했습니다!

malibinYun commented on src/test/kotlin/omok/fixture/BoardFixture.kt:
테스트마다 필요한 보드는 인라인으로 작성해주는 건 어떨까요?
변수명만 보고 어떤 형태의 보드 판을 띄고 있을 지 머리로만 알기는 어려워보여요.
그것을 연상시키는 데 도울 수 있는 방법이 무엇이 있을지도 고민해보면 좋겠네요 :)

또, 보드의 초기 값을 placeStone 함수를 불러 채워주기보다,
생성할 때 초기 값을 넣을 수 있도록 구성해보는 건 어떨까요?

moondev03 replied:
기존에 작성해놓은 Fixture들을 모두 인라인으로 수정했습니다.
대신 TestBoard를 생성에 도움을 주는 함수를 추가했습니다.

또한 Board에 placeStone을 호출하여 생성하는 부분에서는 착수할 때마다 여러 로직이 포함되어 있기 때문에
문제가 발생할 수 있는 것 같아 points를 초기값으로 받을 수 있도록 수정했습니다.

malibinYun reviewed (changes requested):
1, 2단계 미션 고생 많으셨어요 :)
여러 생각해볼만한 곳에 코멘트 남겨두었어요.
충분히 고민해보신 시간 가져보세요 :)
피드백 반영 후 다시 리뷰 요청 부탁드려요

moondev03 committed:
refactor: Board 개선

- points 자료 구조 변경
- Position 제거 후 Point에 역할 부여
- StoneColor 제거

moondev03 committed:
refactor: OmokGame에서 View 대신 Listener 사용하도록 변경

moondev03 committed:
refactor: Adapter -> Converter 네이밍 수정

moondev03 committed:
refactor: Rules object -> class 변경

moondev03 committed:
refactor: PointState 색상 변경 기능 수정

moondev03 committed:
refactor: Retry 로직에 shouldRetry 조건 추가

moondev03 committed:
refactor: Rule를 관리하는 객체 추가

- 게임 & 보드 전체 구조 변경

moondev03 committed:
refactor: Board 역할 분리

- BoardPoints
- BoardSize

moondev03 committed:
refactor: Board 역할 분리

- BoardPoints
- BoardSize

moondev03 committed:
refactor: OmokGame 개선

- 가독성 개선

moondev03 committed:
test: FiveInRowRule 가독성 개선

moondev03 commented:
## OmokGame - View Listener
OmokGame에 View Interface를 받도록 하는 대신 Listener를 통해 Game이 동작할 수 있도록 수정했습니다.

<br>

## OmokGameJudge
오목 게임에서 판단해야 하는 승리 조건, 예외 조건을 관리하고 검사하는 객체를 추가했습니다.
해당 객체는 OmokGame의 생성자로 받도록 했습니다.

<br>

## Board 역할 분리
기존 Board는 Points 생성 및 관리, Size 관리, 착수 등 너무 많은 역할을 수행하고 있다고 생각되어서 역할을 분리해봤습니다.
- BoardSize - 바둑판의 크기를 관리, 검증하는 역할
- BoardPoints - 바둑판의 칸들을 관리하고 각 칸의 상태를 관리하는 역할
- Board - 바둑돌 착수에 대한 결과를 반환하는 역할

malibinYun commented on src/main/kotlin/omok/controller/OmokGameListener.kt:
```diff
@@ -0,0 +1,17 @@
+import omok.model.board.Point
+import omok.model.board.PointState
+
+interface OmokGameListener {
+    fun onStartGame()
+
+    fun onRequestPosition(previousPoint: Pair<Point?, PointState>): Point
+
```
onStartGame 까지는 이름이 적절해보이는데, onRequestPosition은 정말로 리스너의 형태인가요?
이름을 리스너로 지은 이유가 궁금하네요.

moondev03 replied:
지금 보니 onRequestPosition은 듣는 역할만 하는 것은 아닌 것 같네요.
EventHandler로 네이밍만 수정 or onRequestPosition의 역할만 분리
둘 중 더 나은 방법은 무엇인 지 고민해보겠습니다.

moondev03 replied:
UI 출력 방식이 바뀌거나 입력 방식이 다양해질 가능성이 있다면 분리하는 것이 유지보수 측면에서 더 안전할 수도 있을 것 같습니다.
다만, 현재는 분리할 필요성을 느끼지 못해서 `Listener`에서 `Handler`로 네이밍만 수정했습니다.

malibinYun commented on src/main/kotlin/omok/model/OmokGame.kt:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private fun playTurns(board: Board): Boolean =
+        retryOnException(
+            action = {
+                val point = getNextPoint()
+                val placeResult = board.placeStone(point, currentStoneColor)
+                handlePlaceResult(placeResult, board)
+            },
+            shouldRetry = { it },
+        )
```
이 코드를 다른 아무 개발자에게 보여주면, 어떤식으로 동작하는지, 람다가 무엇을 반환해야 어떤 동작을하는지 유추하는 게 쉬울까요?

moondev03 replied:
PlaceResult 리팩터링을 통해 종료 여부를 확인할 수 있게 되었습니다.
아래와 같이 수정해봤습니다!

```kotlin
private fun playTurns(board: Board) {
    do {
        val point = getNextPoint()
        val placeResult = board.placeStone(point, currentStoneColor)
        handlePlaceResult(placeResult, board)
    } while (placeResult is OnGoing)
}
```

malibinYun commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,24 @@
+    val points: BoardPoints,
+    private val judge: OmokRuleJudge,
+) {
+    fun findPointState(point: Point): PointState? = points.getState(point)
+
```
저는 아직도 Point와 PointState가 뭐가 다른지 모르겠어요.
좌표의 상태인데.. 색깔이있네요?
"검정색인 좌표의 상태"는 저에게는 너무 어려운 말이네요

moondev03 replied:
PointState를 제거하고 StoneColor만을 나타내도록 수정했습니다.

[c28427b](https://github.com/woowacourse/kotlin-omok/pull/108/commits/c28427bf554719c8dca7a39b1448708f0526438a)

malibinYun commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,24 @@
+        val state = findPointState(point) ?: return PlaceStoneResult.Failure.InvalidPoint
+        if (state != PointState.OPEN) return PlaceStoneResult.Failure.AlreadyPlaced
+
+        if (!judge.validate(this, point, pointState)) return PlaceStoneResult.Failure.Closed
+        points.update(point, pointState)
```
OmokRuleJudge와 Board는 양방향 의존성을 가지네요.
어떤 문제가 생길 수 있을까요?

moondev03 replied:
양방향으로 모델이 의존하게 된다면 복잡성이 증가하고 유연성이 감소하는 문제가 발생할 것 같습니다.

---
양방향 의존성을 고려하여 최대한 문제를 최소화하려고 했습니다.

현재의 `RuleValidator`는 오목뿐만 아니라 바둑알 및 바둑판을 사용하는 다른 게임에도 적용할 수 있도록 구현했습니다
위반 규칙, 승리 규칙, 각 규칙마다의 타겟을 추가할 수 있습니다.

`Board`에서는 위에서 정의된 validator를 사용하여 타겟에 맞춰 추상화된 규칙들을 검증하고, 승리 여부를 판단합니다.

malibinYun commented on src/main/kotlin/omok/model/board/Board.kt:
Board가 맡아볼 수 있는 일을 다른 객체가 대신 해주고 있다는 느낌이 강하게 들어요.
이 객체가 할 수 있는 일을 다른 객체로부터 옮겨보는 건 어떨까요?

moondev03 replied:
`BoardPoints`: `Map<Point, StoneColor?>`는 충분히 `Board`안에서 역할을 수행 가능하다고 판단해서
다시 `Board`에서 `Points`를 관리하도록 수정했습니다.

[6da7843](https://github.com/woowacourse/kotlin-omok/pull/108/commits/6da78431f9a5d53b9335f0c2f556ce0f1cd07d07)

malibinYun commented on src/main/kotlin/omok/model/board/PlaceStoneResult.kt:
성공과 실패로 두 단계를 나눈 큰 이유가 있었나요?
그냥 글로만 봤을 때, 이미 있는 곳에 두는 건 "실패"한 결과인가요?

너무 에러와 정상 케이스라는 개념에 갇혀있지 않았으면 해요 :)

moondev03 replied:
단순히 `착수 성공`, `착수 실패`를 생각하고 작성했습니다.
에러와 정상 케이스로 생각하지 않고 `게임이 종료되는 경우`와 `종료되지 않는 경우`로 분리해보겠습니다!

[0bccc17](https://github.com/woowacourse/kotlin-omok/pull/108/commits/0bccc174c3422c5d9614177e74f2ee058362b72a)

malibinYun commented on src/main/kotlin/omok/model/rule/OmokConverter.kt:
```diff
@@ -0,0 +1,26 @@
+import omok.model.board.Point
+import omok.model.board.PointState
+import omok.model.rule.lib.ForbiddenMoveRule
+
+object OmokConverter {
```
오목 규칙을 다른 라이브러리로 교체하거나, 여러개를 동시에 사용해야한다면, 이 객체는 어떤식으로 변화하게 될까요?

moondev03 replied:
Converter 제거 후 Adapter 추가했습니다!

[0a939a8](https://github.com/woowacourse/kotlin-omok/pull/108/commits/0a939a88e1fe938bb67c97151f809cffa44d3aef)

malibinYun commented on src/main/kotlin/omok/model/rule/OmokRuleJudge.kt:
```diff
@@ -0,0 +1,70 @@
+        winningRules[rule] = listOf(PointState.BLACK, PointState.WHITE)
+    }
+
+    fun applyRenjuRule() {
+        rules +=
+            mapOf(
+                DoubleThreeMoveRule() to listOf(PointState.BLACK),
+                DoubleFourMoveRule() to listOf(PointState.BLACK),
+                OverlineRule() to listOf(PointState.BLACK),
+            )
```
https://github.com/woowacourse/kotlin-omok/pull/108#discussion_r2006951313

여전히 유효해요.
게임 규칙이 게임 마다 바뀌어야한다면 유연하게 대처가 가능할까요?
모든 룰에 대해 1개씩 apply 함수를 만들어내는 게 유일한 해결법일까요?

moondev03 replied:
조금 더 유연하게 대처 가능하도록 외부에서 추가하도록 수정했습니다.

[0a939a8](https://github.com/woowacourse/kotlin-omok/pull/108/commits/0a939a88e1fe938bb67c97151f809cffa44d3aef)

malibinYun commented on src/test/kotlin/omok/fixture/BoardFixture.kt:
```diff
@@ -0,0 +1,22 @@
+import omok.model.rule.OmokRuleJudge
+import omok.model.rule.count.FiveInRowRule
+
+fun generateTestBoardFixture(
+    points: List<Point>,
+    state: PointState,
+): Board {
+    val judge =
+        OmokRuleJudge().apply {
+            applyRenjuRule()
+            applyWinningRule(FiveInRowRule())
+        }
+    val boardPoints = BoardPoints(BoardSize.DEFAULT, points.associateWith { state })
+    return Board(boardPoints, judge)
```
인라인의 의미는 분리된 코드를 해당 위치에 전부 옮기는 의미예요.
Board의 픽스처인데 불구하고, 오목 룰에 대한 코드를 실행시켜야만 하네요.
모든 테스트가 이 룰에 의존을 강하게 가지겠어요. 조금이라도 다른 룰을 다른 테스트에서 적용해야한다면,
그떄마다 그런 함수를 무한정 만들어내는 게 유일한 해결법일까요?

또, 여전히 생성 코드만을 보고는 어떤 형태인지 머리로 그려내기는 어려워요.
주석을 활용해서 시각적인 도움을 주는 그림을 그려보세요.

moondev03 replied:
테스트 코드 개선 진행했습니다.
3-3, 4-4 금수, 오목 같은 경우는 주석으로 바둑판 그림 추가했습니다!

malibinYun reviewed (changes requested):
피드백 반영 고생 많으셨어요.
뭔가 갑자기 제가 리뷰를 보는 다른 크루의 것과 비슷해져서 헷갈렷네요.
스스로의 것으로 만들어내는 연습을 해보셔도 좋겠어요.
이번 미션은 화요일~수요일 사이에 머지하도록 할게요. 다음 미션도 중요하니까요.
피드백 반영 후 다시 리뷰 요청 해주세요!

moondev03 committed:
refactor: PointState -> StoneColor

- 기존 PointState를 돌의 색상만 나타내도록 변경
- OPEN을 제거하고 Board의 각 Point에는 StoneColor 혹은 null을 가지도록 수정

moondev03 committed:
refactor: Board 개선

- BoardPoints 제거 후 Board로 역할 이전

moondev03 committed:
refactor: PlaceStoneResult 개선

- PlaceStoneResult Interface로 변경
- Finished / OnGoing으로 종료되는 경우와 아닌 경우를 분리

moondev03 committed:
refactor: Default BoardSize 제거

moondev03 committed:
refactor: RuleValidator 추가

- Converter 제거
- Adapter 추가
- 테스트 코드 정리

moondev03 committed:
refactor: View Listener -> Handler로 네이밍 변경

moondev03 committed:
refactor: Retry 가독성 개선

moondev03 committed:
refactor: OmokRule -> GameRule

moondev03 committed:
refactor: ViewExtension 가독성 개선

moondev03 committed:
refactor: Retry 수정

moondev03 commented:
## RuleValidator
기존 `OmokRuleJudge` 클래스로 현재는 `GameRule` Interface를 상속받은 룰들을 통해
위반조건, 승리조건, 각 룰에 타겟을 추가할 수 있도록 했습니다.

또한 `Board`도 바둑판과 바둑알을 사용하는 다른 게임에서도 어느 정도 사용될 수 있도록 RuleValidator를 주입받을 수 있도록 했습니다.

<br>

## PlaceStoneResult
기존의 `PlaceStoneResult`를 `성공`과 `실패`로 분리했었는데, 이를 Interface로 두고 `OnGoing`과 `Finished`로 나누어 구현해봤습니다.

이렇게 수정함으로써 여러 함수에서 반환되던 `boolean` 값들을 제거할 수 있었고, `retry` 로직도 개선할 수 있었습니다.

malibinYun commented on src/main/kotlin/omok/controller/OmokGameHandler.kt:
https://github.com/woowacourse/kotlin-omok/pull/108#discussion_r2009361248

오목 게임 핸들러... 게임을 핸들링한다...
뭔가 게임을 전체적으로 제어할 것 같은 이름인걸요?

또, onRequestPosition은 메서드명으로 적합할지 고민해보세요.

malibinYun commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,71 @@
+    size: BoardSize,
+    points: Map<Point, StoneColor?> = emptyMap(),
+    private val validator: RuleValidator,
+) {
+    private val _points: MutableMap<Point, StoneColor?> = points.toMutableMap()
```
null 로 빈 칸을 나타내지 않고, enum으로 나타내보는 건 어떨까요?
enum을 자연스럽게 이름이 바뀌겠네요 :)

moondev03 replied:
null 대신 NONE을 추가했습니다.

malibinYun commented on src/main/kotlin/omok/model/rule/count/GameCountRuleAdapter.kt:
어뎁터 패턴을 잘 활용하고 있는지 고민해보시고, 개선해보시면 좋겠어요.
돌아가서 현재 어뎁터 패턴을 사용하는 것이 적절한지도 고민해보시면 좋겠네요.

malibinYun commented on src/main/kotlin/omok/model/board/result/PlaceStoneResult.kt:
```diff
@@ -0,0 +1,3 @@
+package omok.model.board.result
+
+interface PlaceStoneResult
```
PlaceStoneResult sealed 로 묶지 않은 이유는 무엇인가요?

moondev03 replied:
`sealed class` 아래에 다시 `sealed class`로 `OnGoing`과 `Finished`를 두는 것보다, `interface`를 상속받은 `OnGoing`과 `Finished`를 두는 것이 확장성 측면에서 더 낫다고 생각했습니다. 

또한 상태가 늘어난다면 `PlaceStoneResult`가 너무 커질 수 있어, `interface`를 사용하는 방식이 더 적합하다고 판단했습니다.

malibinYun commented on src/main/kotlin/omok/model/rule/RuleValidator.kt:
```diff
@@ -0,0 +1,50 @@
+import omok.model.board.Point
+import omok.model.board.StoneColor
+
+class RuleValidator {
+    private val violationRules: MutableMap<GameRule, List<StoneColor>> = mutableMapOf()
+    private val winningRules: MutableMap<GameRule, List<StoneColor>> = mutableMapOf()
+
+    fun addWinningRule(
+        rule: GameRule,
+        targetColors: List<StoneColor>,
+    ) {
+        winningRules[rule] = targetColors
+    }
+
+    fun addViolationRule(
+        rule: GameRule,
+        targetColors: List<StoneColor>,
+    ) {
+        violationRules[rule] = targetColors
+    }
+
```
이 객체를 생성할 때 초기 값을 넣을 수 있도록 구성해보는 건 어떨까요?

malibinYun commented on src/main/kotlin/omok/model/rule/GameRule.kt:
```diff
@@ -0,0 +1,11 @@
+import omok.model.board.Board
+import omok.model.board.Point
+
+interface GameRule {
+    fun validateMove(
+        board: Board,
+        previousPoint: Point,
+    ): Boolean
+}
```
이해가 잘 안돼서 그러는데, 보드와 이전 좌표가 필요한 이유는 무엇인가요?
함수 시그니처만 보고는 어떤 일을 하는지 짐작이 잘 가지 않네요.

moondev03 replied:
규칙 위반인지 검증할 좌표를 나타내기 위해
previousPoint -> validationPoint로 수정했습니다!

malibinYun commented on src/main/kotlin/omok/model/rule/count/OmokCountRule.kt:
OmokCountRule 과 ForbiddenMoveRule은 착수 규칙이란 점에서 맥락이 같아보여요.
둘을 하나의 룰로 통합해보는 건 어떨까요?

malibinYun commented on src/test/kotlin/omok/model/board/BoardTest.kt:
```diff
@@ -0,0 +1,132 @@
+        val size = BoardSize(15)
+        val board = Board(size, judge = RuleValidator())
+
+        assertThat(board.size).isEqualTo(15)
+    }
+
+    @Test
+    fun `초기 보드는 모든 점이 null이다`() {
+        createBoard().points.forEach { (_, state) ->
+            assertEquals(null, state)
+        }
+    }
```
assertThat 과 assertEquals를 혼용해서 쓰는 이유는 무엇인가요?
둘의 차이점은 무엇일까요?

moondev03 replied:
같은 기능이라 생각해서 작성할 때 작성할 때 다르다고 인식하지 못했던 것 같습니다.

차이점은 가독성이 AssertJ이 좋은 것 같습니다.
junit을 사용하면 비교값이 무엇이 실제값이고 무엇이 예측값인지 한눈에 파악하기 어려운 것 같습니다.

추가로 AssertJ는 메소드 체이닝도 가능하다는 장점도 있는 것 같습니다!

malibinYun commented on src/test/kotlin/omok/model/board/BoardTest.kt:
```diff
@@ -0,0 +1,132 @@
+        val initialPoints = mapOf(Point(1, 1) to StoneColor.BLACK, Point(2, 2) to StoneColor.WHITE)
+        val customBoard = Board(BoardSize(15), initialPoints, validator = RuleValidator())
+
+        assertEquals(StoneColor.BLACK, customBoard.findStoneColor(Point(1, 1)))
+        assertEquals(StoneColor.WHITE, customBoard.findStoneColor(Point(2, 2)))
+        assertEquals(null, customBoard.findStoneColor(Point(3, 3)))
```
assertAll로 감싸는 것과 아닌 것은 어떤 차이가 있을까요?

moondev03 replied:
`assertAll`: 여러 검증을 하고, 실패한 부분을 모두 보고하고 싶은 경우.
`assertSoftly`: 여러 검증을 하면서 실패를 중단하지 않고, 모든 검증을 끝낸 후 결과를 보고하고 싶은 경우.

지금은 먼저 테스트한 케이스가 실패하는 경우 바로 종료되기 때문에 다른 케이스의 실패 여부를 확인하기 어렵다는 문제가 있는 것 같습니다!

malibinYun reviewed (approved):
1, 2 단계 미션 고생 많으셨어요.
이제 안드로이드로 넘어가면서 짠 코드를 잘 활용할 수 있을지 검증해보는 시간이 왔네요 :)
다음 단계를 위해 이만 머지하겠습니다.
피드백은 다음 미션에 반영해주세요 :)
